### ISSUE TITLE
Erroneous error when subclassing Sequence

### ISSUE THREAD

**Issue by leandropls** [python/mypy#4108]
Suppose the following snippet:

    class FloatSequence(Sequence[float]):
        __slots__ = ('data',)
        data: Tuple[float, ...]
    
        def __init__(self, more: Iterable[float]) -> None:
            if isinstance(more, FloatSequence):
                self.data = cast(FloatSequence, more).data
            else:
                self.data = tuple(float(x) for x in more)
    
        @overload
        def __getitem__(self, index: int) -> float:
            pass
    
        @overload
        def __getitem__(self, index: slice) -> 'FloatSequence':
            pass
    
        def __getitem__(self,
                        index: Union[int, slice]) -> Union[float, 'FloatSequence']:
            if isinstance(index, slice):
                return FloatSequence(self.data[index])
            else:
                return self.data[index]
    
        def __len__(self) -> int:
            return len(self.data)

Checking it with mypy yields:

    $ mypy scratch_20.py 
    scratch_20.py:24: error: Signature of "__getitem__" incompatible with supertype "Sequence"

Removing the type annotations of the implementation of `__getitem__` fixes the error -- though, from the docs, annotating the implementation seems to be the right way to do it.

**Comment by gvanrossum** [comment]
Interesting, if you replace the return type of the implementation with `Any`, it also passes. It seems mypy still type-checks the implementation against `Sequence.__getitem__`. Agreed this is a bug, since it should just be checked against the overloads, and the overloads should be checked against the base class.

**Comment by leandropls** [comment]
Just replaced the return types of the 2nd overload and the implementation of `__getitem__` to better reflect what I was assuming should work: instead of `Sequence[float]` it's now annotated as `FloatSequence`.

**Comment by ilevkivskyi** [comment]
The example above now passes, it looks like the issue is fixed now.

**Comment by randolf-scholz** [comment]
This bug still occurs when one uses `@singledispatchmethod` instead of `@overload` (tested with python 3.9.7 and mypy 0.941)

```python
from collections.abc import Sequence, Iterable
from typing import Union, cast, overload
from functools import singledispatchmethod


class FloatSequence(Sequence[float]):
    __slots__ = ("data",)
    data: tuple[float, ...]

    def __init__(self, more: Iterable[float]) -> None:
        if isinstance(more, FloatSequence):
            self.data = cast(FloatSequence, more).data
        else:
            self.data = tuple(float(x) for x in more)

    @overload
    def __getitem__(self, index: int) -> float:
        pass

    @overload
    def __getitem__(self, index: slice) -> "FloatSequence":
        pass

    def __getitem__(self, index: Union[int, slice]) -> Union[float, "FloatSequence"]:
        if isinstance(index, slice):
            return FloatSequence(self.data[index])
        else:
            return self.data[index]

    def __len__(self) -> int:
        return len(self.data)


class IntSequence(Sequence[int]):
    __slots__ = ("data",)
    data: tuple[float, ...]

    def __init__(self, more: Iterable[float]) -> None:
        if isinstance(more, IntSequence):
            self.data = cast(IntSequence, more).data
        else:
            self.data = tuple(float(x) for x in more)

    @singledispatchmethod
    def __getitem__(  # error: Signature of "__getitem__" incompatible with supertype "Sequence"
        self, index: Union[int, slice]
    ) -> Union[float, "IntSequence"]:
        raise TypeError("unsupported index type")

    @__getitem__.register
    def _(self, index: int) -> float:
        return self.data[index]

    @__getitem__.register
    def _(self, index: slice) -> "IntSequence":
        return IntSequence(self.data[index])

    def __len__(self) -> int:
        return len(self.data)
```
